R/lm_lin.R

```diff
@@ -82,21 +82,15 @@ lm_lin <- function(formula,
       # create dummies for non-factor treatment variable
 
       # Drop out first group if there is an intercept
-      vals <- sort(unique(treatment))[if(has_intercept) {-1} else {TRUE}]
+      vals <- sort(unique(treatment))
+      if (has_intercept) vals <- vals[-1]
+
       n_treats <- length(vals)
-      # TODO warn if too many values?
-
-      treatment_mat <- matrix(
-        NA,
-        nrow = n,
-        ncol = n_treats,
-        dimnames = list(NULL,
-                        paste0(colnames(design_matrix)[treat_col], vals))
-      )
+      # TODO warn if too many values? (ie. if there are as many treatments as observations, maybe we should warn)
 
-      for (i in 1:n_treats) {
-        treatment_mat[, i] <- as.numeric(treatment == vals[i])
-      }
+      # Create matrix of dummies
+      treatment_mat <- outer(drop(treatment), vals, function(x, y) as.numeric(x == y))
```

Comment:
```
treatment_mat <- 1.0 * outer(X, Y, `==`)
```

Reply:
`outer` can also set the names correctly, as in:
```
> outer(setNames(nm=LETTERS[1:3]), setNames(nm=letters[1:3]), `==`)
      a     b     c
A FALSE FALSE FALSE
B FALSE FALSE FALSE
C FALSE FALSE FALSE
```

so you could do the naming a little earlier
```
names(vals) <- paste0(colnames(design_matrix)[treat_col], vals)
outer(drop(treatment), vals)
```

and at this point, there's enough logic around vals to justify a helper function - `get_vals_from_treat_matrix()` or similiar.

Reply:
Is there any reason to prefer 1.0 to as.numeric?